Model/DutyCalculator.php

```diff
@@ -116,84 +116,333 @@ public function __construct(
+     * @param Magento\Quote\Api\Data\AddressInterface $address
+     */
+    public function handleCaseWhereDutyIsAlreadyRetrieved($apply, $quote, $address)
+    {
+        if ($apply) {//user indicated that (s)he wants to apply duty and tax during calculation of billing
+            $quote->setReachDuty($this->checkoutSession->getReachDuty());
+            $this->response->setDuty($this->checkoutSession->getReachDuty());
+            $this->handleTaxApplicability($address, $apply);
+            $this->_logger->debug('User wants to apply the D&T. Value of apply ::' . $apply);
+            $quote->setBaseReachDuty($this->checkoutSession->getBaseReachDuty());
+            $quote->setReachDuty($this->checkoutSession->getReachDuty());
+            $quote->setDhlQuoteId($this->checkoutSession->getDhlQuoteId());
+            $quote->setDhlBreakdown($this->checkoutSession->getDhlBreakdown());
+            $this->_logger->debug($quote->getReachDuty());
+        }
+        else {//user indicated that (s)he does not  want to apply duty and tax during calculation of billing
+            $quote->setReachDuty(0);
+            //should we reset DhlQuoteId and DhlBreakdown too so as not to break any reporting?
+            //I am assuming that we do
+            $quote->setBaseReachDuty(0);
+            $quote->setDhlQuoteId('');
+            $quote->setDhlBreakdown('');
+            $this->_logger->debug('User does not want to apply the D&T. Value of apply ::' . $apply);
+        }
+
+        $quote->save();
+        $this->response->setDuty($this->checkoutSession->getReachDuty());
+    }
+
+    /** Adjust returned duty value from DHL and additionally keep track of different data that came back from
+     * DHL API call and saves proper state of quote so that the information is available beyond this page/area
+     * @param float $duty
+     * @param Magento\Quote\Api\Data\AddressInterface $address
+     * @param bool $apply
+     * @param  Magento\Framework\App\Response $response
+     */
+    public function fillOutQuoteAndSessionUsingFeeReturned($duty, $address, $apply, $response)
+    {
+        //dealing with whether duty and tax related checkbox is selected or not
+        //or whether applying duty is a must for that country or not (a setting in magento admin
+        //panel)
+        //filling out both Quote and session state as appropriate
+        //DHL quoteID is already saved into session; so not doing it here again
+        $quote = $this->checkoutSession->getQuote();
+        $duty_adjusted = $this->priceCurrency->round($duty); //copied over pre-existing code
+        //but are we supposed to round always?
+        //should not that be based on corresponding admin setting?
+        $this->checkoutSession->setReachDuty($duty_adjusted);
+
+        if ($apply || !$this->getIsOptional($address->getCountryId())) {
+            //checkbox selection was duty should be applied
+            //or applying duty is a must for that country
```

Comment:
I didn't think any country had applying duty as a must, could you help me if I missed this requirement in the DHL API spec or somewhere else?

Reply:
*Just copied over/reused this logic from pre-existing code. This is something from default admin config around DHL I think (which affects what behavior we see on store) which is one part of our conversation today as well. I have never seen that part fully working for me. The team indicated during yesterday's SU that I should be talking to you to have some clarity. Thanks.

Reply:
created a ticket to check requirements: https://withreach.atlassian.net/browse/MAG-77